### PR TITLE
Add support for Bson documents to auth custom functions.

### PR DESCRIPTION
**Pull request by edualonso**
Closes:
https://github.com/realm/realm-object-store/issues/1040

The current implementation receives a string payload, which is outdated. It should be a BsonDocument instead.

### CODE REVIEW

File: src/sync/app_credentials.cpp [hunk 1]
```diff
@@ -16,9 +16,11 @@
 //
 ////////////////////////////////////////////////////////////////////////////
 
+#include "util.hpp"
```
Comment (jsflax) [inline src/sync/app_credentials.cpp:19]: Is this includes and `<string>` needed?

Comment (edualonso) [reply src/sync/app_credentials.cpp:19]: `<string>` isn't (probably leftover from some testing I did), but `util.hpp` is needed for parsing the credentials.

Comment (jsflax) [reply src/sync/app_credentials.cpp:19]: This should be `#include "util/bson/bson.hpp"`.

File: src/sync/app_credentials.hpp [hunk 2]
```diff
@@ -21,6 +21,7 @@
 
 #include <functional>
 #include <string>
+#include <object-store/src/util/bson/bson.hpp>
```
Comment (jsflax) [inline src/sync/app_credentials.hpp:24]: This include should not be here and we should be forward declaring BsonDocument.

```cpp
namespace realm {
namespace bson {
class Bson;
template <typename> class IndexedMap;
using BsonDocument = IndexedMap<Bson>;
}
namespace app {
...
```

And then the method signature should be:
```cpp
static AppCredentials function(bson::BsonDocument& payload_json)
```

File: src/sync/app_credentials.hpp [hunk 3]
```diff
@@ -95,8 +96,8 @@ struct AppCredentials {
 
     // Construct and return credentials with the payload.
     // The payload is a MongoDB document as json
-    static AppCredentials function(std::string payload_json);
-    
+    static AppCredentials function(bson::BsonDocument payload);
```
Comment (jsflax) [inline src/sync/app_credentials.hpp:99]: See above comment.

File: src/sync/app_credentials.hpp [hunk 4]
```diff
@@ -95,8 +100,8 @@ struct AppCredentials {
 
     // Construct and return credentials with the payload.
     // The payload is a MongoDB document as json
-    static AppCredentials function(std::string payload_json);
-    
+    static AppCredentials function(bson::BsonDocument& payload);
```
Comment (jsflax) [inline src/sync/app_credentials.hpp:103]: Should make `payload` `const`

**Pull request merged**